### ISSUE TITLE
Build item widgets on demand is called twice

### ISSUE THREAD

**Issue by TCM5** [serenader2014/flutter_carousel_slider#101]
If the same data is loaded in the **classic** vs  **itemBuilder**, building calls are not resulting in the same loaded data. 
This will not result in a functional problem, but if the itemBuilder uses remote API calls to build the item this will make unecessary calls and performance issues.

Example in the classic way:

```
CarouselSlider(
  height: 400.0,
  items: [1,2,3,4,5].map((i) {
    return Builder(
      builder: (BuildContext context) {
        print(i);
        return Container(
          width: MediaQuery.of(context).size.width,
          margin: EdgeInsets.symmetric(horizontal: 5.0),
          decoration: BoxDecoration(
            color: Colors.amber
          ),
          child: Text('text $i', style: TextStyle(fontSize: 16.0),)
        );
      },
    );
  }).toList(),
)
```
On caroussel startup, this code will load items 5,1,2 (yes he is not loading items in ascending way, but that's is another problem). The print ```print(i)``` output is this:

- I/flutter (31887): 5
- I/flutter (31887): 1
- I/flutter (31887): 2

But, changing this code to use item builder like this:
```
CarouselSlider.builder(
                  itemCount: 5,
                  itemBuilder: (BuildContext context, int itemIndex) {
                    print(itemIndex);

                    return Container(
                        width: MediaQuery.of(context).size.width,
                        margin: EdgeInsets.symmetric(horizontal: 5.0),
                        decoration: BoxDecoration(color: Colors.amber),
                        child: Text(
                          'text $itemIndex',
                          style: TextStyle(fontSize: 16.0),
                        ));
                  },
                )
```

On caroussel startup, this code will load items 4,0,1 twice. The print ```print(itemIndex)``` output is this:
- I/flutter (31887): 4
- I/flutter (31887): 0
- I/flutter (31887): 1
- I/flutter (31887): 4
- I/flutter (31887): 0
- I/flutter (31887): 1

I didn't inspect the code for this problem, but this seems a problem related with some setState called twice that will force unecessary reloading (just a guess).

**Comment by TCM5** [comment]
This is suspicious...

``
 builder: (BuildContext context, child) {
            **// on the first render, the pageController.page is null,
            // this is a dirty hack**
            if (widget.pageController.position.minScrollExtent == null ||
                widget.pageController.position.maxScrollExtent == null) {
              Future.delayed(Duration(microseconds: 1), () {
	        **if (this.mounted) {
                  setState(() {});
		}**
              });
              return Container();
            }``

**Comment by serenader2014** [comment]
Hi, sorry for the late response.

Thank you for reporting the issue, I'll look into it.

> This will not result in a functional problem, but if the itemBuilder uses remote API calls to build the item this will make unecessary calls and performance issues.
For this scenario, please take a look at this SO link: https://stackoverflow.com/questions/52249578/how-to-deal-with-unwanted-widget-build

Putting API call in the build method is not recommend.

**Comment by serenader2014** [comment]
Fixed in `v2.0.0`

**Comment by omidraha** [comment]
I may have a related problem.
I'm in `3.0.0`, and `itemBuilder` of `CarouselSlider.builder` in the previous page is  still called again when I navigated  to another new page by `onPageChanged`.
